---
title: "Chart Combinations"
author: "Anamaria Crisan"
date: '2019-01-09'
output: html_document
---


This document demonstrates how to combine and plot different chart types using gevitR. This document assumes that you have already run the "Getting started with gevitR" and the "Simple Chart Types" and that the necessary data as already been loaded into your R workspace.

Another good option for data: https://datadryad.org/bitstream/handle/10255/dryad.111738/ZIKV_with1966_strict_skyline_ctmc.xml?sequence=1
Also exists in microreact

## Loading data
```{r,message=FALSE,warning=FALSE}
devtools::load_all() #temporary once things are done
library(dplyr)
library(ggtree)

#Table data
tab_dat<-input_data(file = system.file("extdata", "ebov_metadata.csv", package = "mincombinr"),dataType = "table")

#Tree data
tree_dat<-input_data(file = system.file("extdata", "ebov_tree.nwk", package = "mincombinr"),dataType = "tree")

#Genomic data
genomic_dat<-input_data(file = system.file("extdata", "ebov_GIN_genomic_FIXED.fasta", package = "mincombinr"),dataType = "dna")

#Shape files
#Shape files require that .shp,.shx,and .prj files at a minimun to be in the same directory
#to add metadata to the shape file, you can also add .dbf files
gin_shape_dat<-input_data(file = system.file("extdata", "gin_admbnda_adm1_ocha_itos.shp", package = "mincombinr"),dataType = "spatial")
lbr_shape_dat<-input_data(file = system.file("extdata", "lbr_admbnda_adm1_ocha.shp", package = "mincombinr"),dataType = "spatial")
sle_shape_dat<-input_data(file = system.file("extdata", "sle_admbnda_adm1_1m_gov_ocha_20161017.shp", package = "mincombinr"),dataType = "spatial")

```

## Chart Combinations

First, let's create a few charts that can be used
```{r individual_charts,message=FALSE,warning=FALSE}
bar_chart<-specify_base(chart_type = "bar",data="tab_dat",x = "country")

phyloTree_chart<-specify_base(chart_type = "phylogenetic tree",data="tree_dat")

scatter_chart<-specify_base(chart_type = "scatter",data="tab_dat",x = "latitude",y="longitude",title = "Ebola Scatter Plot")

map_chart<-specify_base("geographic map",data="tab_dat",lat = "latitude",long = "longitude")
```

### Many Types General
Many types general are essentially when the you just want to put a bunch of plots together and there are no spatial or visual linkages between the plots themselves.

```{r many_types_general,message=FALSE,warning=FALSE}
mg_combo<-specify_combination(combo_type = "many_types_general",base_charts = c("bar_chart","phyloTree_chart","scatter_chart","map_chart"))
plot(mg_combo)
```

### Small Multiples
Small multiple charts are visually linked because the show the same underlying chart type, however they show different aspects of the data.

```{r small_multiples,message=FALSE,warning=FALSE}
#set up the base chart type with all the data that you wish to use
scatter_chart<-specify_base(chart_type = "scatter",data="tab_dat",x = "latitude",y="longitude")

#now specify the small multiple combination
sm_combo_scatter<-specify_combination(combo_type = "small_multiple",base_charts = "scatter_chart",facet_by = "country")
plot(sm_combo_scatter)

sm_combo_bar<-specify_combination(combo_type = "small_multiple",base_charts = "bar_chart",facet_by = "country")
plot(sm_combo_bar)

#let's try a more interesting bar_chart small multiple
bar_chart_alt<-specify_base(chart_type = "bar",data="tab_dat",x = "year",title="All together")
sm_combo_bar_alt<-specify_combination(combo_type = "small_multiple",base_charts = "bar_chart_alt",facet_by = "country")
plot(bar_chart_alt)
plot(sm_combo_bar_alt)

```

Other chart types cannot be easily subsetted. For example, it's not straight forward to make a small multiple of a phylogenetic tree while only showing a subset of the tree, the same is generally true for maps, and networks. Yes there are ways to truely subset them, but it's messy and in the whole network structure matters, so here you get the whole network structure.

In the current implementation of gevitR, there needs to be some tablular data assocaited with non-tabular data in order to be able to understand what should be visualized in the first place.

```{r small_multiples_complex,message=FALSE,warning=FALSE}
#Tree data
tree_dat<-input_data(file = system.file("extdata", "ebov_tree.nwk", package = "mincombinr"),dataType = "tree")

tree_dat_meta<-input_data(file = system.file("extdata", "ebov_tree.nwk", package = "mincombinr"),dataType = "tree",metadata = system.file("extdata", "ebov_metadata.csv", package = "mincombinr"))

tree_chart<-specify_base(chart_type = "phylogenetic tree",data="tree_dat_meta")
plot(tree_chart)

sm_combo_tree<-specify_combination(combo_type = "small_multiple",base_charts = "tree_chart",facet_by = "country")
plot(sm_combo_tree)

```

### Color Linked Combinations

Finally, it could be interesting to link serveral different chart types together by colour. In the above examples, we may want to link the phylogenetic tree with the timeline according to the different countries. For the non-tabular data, it's important to have so associated metadata, otherwise, it is not possible to link information. It is up to the user to establish that two variables are actually linkable by the same variable. Some of the code form the composite combination is borrowed to see if two datasets are even linkable to help with the color linkage.

**First scenario - no metadata provided for the tree**

minCombinR will try to find if there are linkages and if there are EXACT MATCHES will link tabular data as tree's metadata
```{r}
phyloTree_chart<-specify_base(chart_type = "phylogenetic tree",data="tree_dat")
epicurve<-specify_base(chart_type = "histogram",data="tab_dat",x = "month")

color_combo<-specify_combination(combo_type = "color_linked", base_charts = c("phyloTree_chart","map_chart","epicurve"),link_by="country")

plot(color_combo)
```

### Composite combinations

*This example will work without any errors*
```{r composite_no_error,message=FALSE,warning=FALSE}
scatter_chart<-specify_base(chart_type = "scatter",data="tab_dat",x = "month",y="site.id")
scatter_chart_two<-specify_base(chart_type = "scatter",data="tab_dat",x = "country",y="site.id",title="Cases by country")

composite_combo<-specify_combination(combo_type = "composite",
                                     base_charts = c("phyloTree_chart","scatter_chart","scatter_chart_two"))

plot(composite_combo)

```

Example without the phylogenetic tree


*This example will have an error because pie charts cannot be part of composite combinations*
Pie chart should be autonatically removed from the specifications
```{r composite_no_error_two,message=FALSE,warning=FALSE,}

pie_chart<-specify_base(chart_type = "pie",data="tab_dat",x = "country")
composite_combo<-specify_combination(combo_type = "composite",
                                     base_charts = c("phyloTree_chart","scatter_chart","scatter_chart_two","pie_chart"))


plot(composite_combo)

```

```{r}
bar_alt<-specify_base(chart_type = "bar",data="tab_dat",x = "site.id",y="month",rm_x_label=TRUE)

composite_combo<-specify_combination(combo_type = "composite",
                                     base_charts = c("phyloTree_chart","bar_alt","scatter_chart_two"))

plot(composite_combo)

```

*It also works when there isn't a tree invovled.
```{r}
composite_combo<-specify_combination(combo_type = "composite",
                                     base_charts = c("bar_alt","scatter_chart"))

plot(composite_combo)

```

Here's a combination with a genomic map..
```{r,fig.width=10}
#genomic chart
#for illustrative purposes, show fewer positions
diff_seq<-get_diff_pos(genomic_dat)
genome_chart<-specify_base(data = "genomic_dat",chart_type = "alignment",title="Genome Alignment",show_pos=diff_seq[1:20])


#timeline, with some fake end_dates
#using the existing tabular data
#using the existing tabular data
time_dat<-tab_dat@data[[1]]
time_dat$collection_date<-as.Date(time_dat$collection_date)

#let's add some end dates to keep it interesting
time_dat$collection_date_end<-time_dat$collection_date + sample(10:30,nrow(time_dat),replace = TRUE)

time_dat$collection_date_end<-sapply(as.character(time_dat$collection_date_end),function(x){
   if(runif(1)>0.9)
     return(x)
    return(NA)
})

time_dat<-dplyr::filter(time_dat,!is.na(collection_date_end))

timeline_chart<-specify_base(chart_type = "timeline"
                             ,data="time_dat",
                             start = "collection_date",
                             end ="collection_date_end", 
                             y = "site.id",
                             title="Timeline")




composite_combo<-specify_combination(combo_type = "composite", base_charts = c("phyloTree_chart","scatter_chart_two","timeline_chart","genome_chart"))
plot(composite_combo)

```

Note that in the above case genomic data is not available for all of the items in the phylogenetic tree. In fact, our source data is for Guinea only. The composite algorithm is able to adjust in instances where one dataset is a perfect subset of the other. It is to the user to ensure this. 
